### DISCUSSION TITLE
PETSC ERROR

### DISCUSSION THREAD

**Discussion by chenyi-maker** [firedrakeproject/firedrake#3115] [Firedrake support] [answered]
I used the docker image when I used firedrake in combination with spyro to compute on a multi-node computer cluster. I can run python run_forward_2d.py in a container without any problems. However, if I run mpiexec -n 10 python run_forward_2d.py in parallel, the following error occurs. Thank you for helping me answer this question.
![6641d9afa388387c65852f6d173be41](https://github.com/firedrakeproject/firedrake/assets/115963126/654341a9-4aec-42fa-bf15-696d0d460a4e)

**Comment by connorjward** [comment]
I am sceptical that Docker will run in a multi-node environment. Have you considered using [singularity](https://github.com/firedrakeproject/firedrake/wiki/singularity)?

**Reply by chenyi-maker** [reply]
Thank you for your answer. My muti-node computer cluster operation is use singularity. But, when I run the .py script in container occurs PETSC ERROR. You have some good idea to slove the problem?

**Reply by connorjward** [reply]
@JDBetteridge is the expert on singularity. Can you share the singularity configuration that you are using? It would also be useful to see the result of running your script with the command line flag `-malloc_debug`.

**Comment by chenyi-maker** [comment]
This is the task script I submitted to the computer cluster. I ran it with the -malloc_debug parameter to resolve the PETSC ERROR. However, after I submitted the script, I encountered the following problem.
![8625cc8892de2792384ff7c2f66186f](https://github.com/firedrakeproject/firedrake/assets/115963126/8dedc8ef-f70b-4dc0-aff9-bb6c5373f798)
![91004d21921b6ff81c35eac89c4616c](https://github.com/firedrakeproject/firedrake/assets/115963126/7cc41e33-b7a0-47e4-875b-9807cca3da67)

**Reply by JDBetteridge** [reply]
This looks like an issue with Singularity. 

Do you not need to load Singularity as a module? Also we have to set a whole load of environment variables to pass through to the container.
Take another look [here,](https://github.com/firedrakeproject/firedrake/wiki/singularity) but it is also worth checking out your clusters documentation and asking for support from the cluster team.It will be much easier for them to reproduce this issue

**Reply by chenyi-maker** [reply, marked as answer]
I really appreciate your help. Based on the information you provided. I succeeded in solving the problem. This is my modified running script for everyone to meet the same problem to do a reference.
![e1994875fcbe857369b000c3d017894](https://github.com/firedrakeproject/firedrake/assets/115963126/50ef1ca8-5a99-4b69-acf7-2b996d84d9cd)

**Reply by JDBetteridge** [reply]
Glad you got everything fixed :slightly_smiling_face:
